[NRBF] Address issues discovered by Threat Model

adamsitnik opened the pull request:
fixes #106644

adamsitnik committed:
introduce ArrayRecord.TotalElementsCount

adamsitnik committed:
remove recursion

adamsitnik committed:
don't allocate the Stack unless it's necessary

adamsitnik committed:
use Queue (both are fine as the order does not really matter here)

adamsitnik committed:
do not include invalid Type or Assembly names in the exception messages, as it's most likely corrupted/tampered/malicious data and could be used as a vector of attack.

adamsitnik commented on src/libraries/System.Formats.Nrbf/src/Resources/Strings.resx:
Based on what we have discussed and what @bartonjs wrote here: https://github.com/dotnet/runtime/pull/103713#discussion_r1717378635

I believe the type names and assembly names should not be provided in the exception messages.

adamsitnik commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/BinaryArrayRecord.cs:
```diff
@@ -27,19 +28,96 @@ internal sealed class BinaryArrayRecord : ArrayRecord
+    private static long GetJaggedArrayTotalElementsCount(BinaryArrayRecord jaggedArrayRecord)
+    {
+        long result = 0;
+        Queue<BinaryArrayRecord>? jaggedArrayRecords = null;
```
The decoder does not perform any recursion when decoding the jagged array. So does this method, but at a cost of potential `Queue` allocation.

adamsitnik committed:
It is possible to have binary array records have an element type of array without being marked as jagged

GrabYourPitchforks commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/BinaryArrayRecord.cs:
```diff
@@ -157,6 +177,65 @@ internal static ArrayRecord Decode(BinaryReader reader, RecordMap recordMap, Pay
+                    case SerializationRecordType.ObjectNullMultiple256:
+                    case SerializationRecordType.ObjectNullMultiple:
+                        // Null Records nested inside jagged array do not increase total elements count.
+                        // Example: "int[][] input = [[1, 2, 3], null]" is just 3 elements in total.
```
Is excluding nulls desirable? (See chat.)

adamsitnik replied:
I've added a comment that explains the null nuances. PTAL

adamsitnik committed:
address feedback:

- rename TotalElementsCount to FlattenedLength
- Ensure lack of recursive call
- add a comment explaining null nuances

adamsitnik committed:
improve exception messages:

- don't include member name in the exception message
- remove unused resource
- make the argument int rather than an object to make it clear it's never a string

adamsitnik commented on src/libraries/System.Formats.Nrbf/src/Resources/Strings.resx:
```diff
@@ -126,26 +126,23 @@
   <data name="Serialization_UnexpectedNullRecordCount" xml:space="preserve">
     <value>Unexpected Null Record count.</value>
   </data>
-  <data name="Serialization_MaxArrayLength" xml:space="preserve">
```
this resource was not being used for a while

adamsitnik commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/ArrayRecord.cs:
```diff
@@ -46,6 +52,11 @@ private protected ArrayRecord(ArrayInfo arrayInfo)
 
+    internal bool IsJagged
+        => ArrayInfo.ArrayType == BinaryArrayType.Jagged
+        // It is possible to have binary array records have an element type of array without being marked as jagged.
+        || TypeName.GetElementType().IsArray;
```
@JeremyKuhne this handles the scenario you have mentioned offline (there is also a test for that). Thank you again for pointing this out!

adamsitnik commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/BinaryArrayRecord.cs:
```diff
@@ -157,6 +177,65 @@ internal static ArrayRecord Decode(BinaryReader reader, RecordMap recordMap, Pay
+                        }
+                        else
+                        {
+                            result += nestedArrayRecord.TotalElementsCount;
```
I've added a `Debug.Assert` with a message to ensure it does not get called for `BinaryArrayRecord`s

adamsitnik commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/BinaryArrayRecord.cs:
```diff
@@ -157,6 +177,68 @@ internal static ArrayRecord Decode(BinaryReader reader, RecordMap recordMap, Pay
+                        else
+                        {
+                            Debug.Assert(nestedArrayRecord is not BinaryArrayRecord, "Ensure lack of recursive call");
+                            result += nestedArrayRecord.FlattenedLength;
```
@GrabYourPitchforks I was wondering whether I should use `checked` here, because in theory somebody could create a payload that would represent a very nested array with total elements count `> long.MaxValue`. Thoughts?

adamsitnik committed:
add checked to ensure no long overflow for very unusual payloads

adamsitnik committed:
FlattenedLength must include nulls

bartonjs commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/BinaryArrayRecord.cs:
```diff
@@ -157,6 +177,74 @@ internal static ArrayRecord Decode(BinaryReader reader, RecordMap recordMap, Pay
+                        break;
+                    case SerializationRecordType.ObjectNull:
+                    case SerializationRecordType.ObjectNullMultiple256:
+                    case SerializationRecordType.ObjectNullMultiple:
+                        // All nulls need to be included, as it's another form of possible attack.
+                        checked
+                        {
+                            result += ((NullsRecord)item).NullCount;
+                        }
+                        break;
+                    default:
+                        result++;
+                        break;
+                }
```
Don't the arrays already know their lengths?  I'd expect FlattenedLength to be a very simple

```C#
long length = here.Length;

foreach (element in array)
{
    if (element.IsArray)
    {
        length += element.FlattenedLength;
    }
}

return length;
```

(or the queue variant)

And that approach would already have counted the `null` elements.  Ensuring that the number of records matches the length is a different problem (one I feel GetArray tries to solve?).

adamsitnik replied:
@bartonjs  You are right. This suggestion has greatly simplified the code and solved the other problem (arrays not being included themselves)

bartonjs commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/BinaryArrayRecord.cs:
```diff
@@ -157,6 +177,74 @@ internal static ArrayRecord Decode(BinaryReader reader, RecordMap recordMap, Pay
+                    case SerializationRecordType.BinaryArray:
+                    case SerializationRecordType.ArraySinglePrimitive:
+                    case SerializationRecordType.ArraySingleObject:
+                    case SerializationRecordType.ArraySingleString:
+                        ArrayRecord nestedArrayRecord = (ArrayRecord)record;
+                        if (nestedArrayRecord.IsJagged)
+                        {
+                            (jaggedArrayRecords ??= new()).Enqueue((BinaryArrayRecord)nestedArrayRecord);
+                        }
+                        else
+                        {
+                            Debug.Assert(nestedArrayRecord is not BinaryArrayRecord, "Ensure lack of recursive call");
+                            checked
+                            {
+                                // In theory somebody could create a payload that would represent
+                                // a very nested array with total elements count > long.MaxValue.
+                                result += nestedArrayRecord.FlattenedLength;
+                            }
+                        }
+                        break;
```
This isn't counting the arrays themselves.  Assuming that's intentional, you should have a comment at the top of the method (or before this switch) that explains what does, and what doesn't count.


```C#
object[][] objs = new object[][1_000_000];
objs.Fill(Array.Empty<object>());
```

Doesn't feel like it should have a smaller count than


```C#
object[][] objs = new object[][1_000_000];
objs.Fill(null);
```

And right now it feels like the first one says 0, and the second says a million.

bartonjs replied:
FWIW, I can understand not counting the arrays as answering how many `int` values might come out of an `int[][][]`; but then it gets weird when the `null`s get counted, because they're not `int` values.

So I guess there's a bit of "what is this number supposed to mean?".  The comment can explain that, and then the code can be judged against the comment.

adamsitnik replied:
Good point, I've fixed that.

bartonjs commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/BinaryArrayRecord.cs:
```diff
@@ -157,6 +177,74 @@ internal static ArrayRecord Decode(BinaryReader reader, RecordMap recordMap, Pay
+                        }
+                        else
+                        {
+                            Debug.Assert(nestedArrayRecord is not BinaryArrayRecord, "Ensure lack of recursive call");
```
I can read the assert, but I don't understand it.  Why can't a BinaryArrayRecord be nested within another BinaryArrayRecord?  Why would that necessarily mean that a recursive call happened?

adamsitnik replied:
I've eliminated the potential recursive call and added a comment.

adamsitnik committed:
Merge remote-tracking branch 'upstream/main' into totalElementsCount

# Conflicts:
#	src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/RectangularArrayRecord.cs

adamsitnik committed:
address code review feedback: count the arrays themselves

adamsitnik commented:
@MihuBot fuzz NrbfDecoder

buyaa-n reviewed (approved):
LGTM and looks @bartonjs's feedbacks addressed as needed

adamsitnik committed:
Merge remote-tracking branch 'upstream/main' into totalElementsCount

# Conflicts:
#	src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/RectangularArrayRecord.cs

adamsitnik commented:
/ba-g DeadLetter is unrelated
